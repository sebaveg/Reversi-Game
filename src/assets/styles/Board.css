
.Board {
  background-color: #8756ca;
  border: 3px solid white;
  box-shadow: 0 0 15px -5px #fff;
  min-width: 280px;
  margin-top: 1em;
}

.wrapper-btn-undo-redo{
  display: flex;
  justify-content: space-between;
  margin-top: .5em;
}

.wrapper-btn-undo-redo button{
  background-color: #8756ca;
  border: 1px solid #8756ca;
  box-shadow: 0 0 5px -2px #fff;
  color: #fff;
  cursor: pointer;
  font-size: 1.5rem;
  font-weight: bold;
  padding: 5px 20px;
}

.wrapper-btn-undo-redo button:disabled {
  visibility: hidden;
}