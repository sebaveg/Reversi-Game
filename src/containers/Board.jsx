import React from 'react';
import { ActionCreators as UndoActionCreators } from 'redux-undo';
import { connect } from 'react-redux';
import {
  setBoard,
  setAllowedCells,
  addDisksPlayers,
  changeTurn,
} from '../actions';

import BoardLayout from '../components/Board';

const directions = [
  [0, 1], // right
  [0, -1], // left
  [-1, 0], // up
  [1, 0], // down
  [1, 1], // down right
  [1, -1], // down left
  [-1, 1], // up right
  [-1, -1], // up left
];

class Board extends React.Component {
  constructor(props) {
    super(props);
    this.rows = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'];
    this.cols = ['1', '2', '3', '4', '5', '6', '7', '8'];
  }

  componentDidMount() {
    this.initialBoard();
  }

  componentDidUpdate(prevProps) {
    const passTurn = this.props.currentPlayer !== prevProps.currentPlayer
    || this.props.posDiskWhite.length !== prevProps.posDiskWhite.length
    || this.props.posDiskBlack.length !== prevProps.posDiskBlack.length;
    if (passTurn) {
      const newBoard = this.allowedCellsAndCountDisks(this.props.board);
      this.props.setBoard(newBoard); // dispatch action
    }
  }

  diskOponent = () => (this.props.currentPlayer === 'white' ? 'black' : 'white');

  inBoard = (x, y) => ((x >= 0) && (x < 8) && (y >= 0) && (y < 8));

  createBoard() {
    const board = this.rows.slice();
    for (let i = 0; i < 8; i += 1) {
      board[i] = this.cols.map((row) => ({
        id: row + board[i],
        disk: null, // white or black
        allowedCell: false, // says if enable or disable click for put disk
      }));
    }
    return board;
  }

  async initialBoard() {
    // Set initial disks: two white and two black in center
    const board = this.createBoard();
    this.props.posDiskWhite.forEach((pos) => {
      board[pos[0]][pos[1]].disk = 'white'; // White
    });
    this.props.posDiskBlack.forEach((pos) => {
      board[pos[0]][pos[1]].disk = 'black'; // Black
    });
    const newBoard = this.allowedCellsAndCountDisks(board);
    await this.props.setBoard(newBoard); // dispatch action
  }

  // says what cells are enable for clicks in current turn
  allowedCellsAndCountDisks(board) {
    const b = board.slice();
    let arrayCanReverse;
    let cantAllow = 0; let disksWhite = 0; let disksBlack = 0;
    for (let x = 0; x < 8; x++) {
      for (let y = 0; y < 8; y++) {
        // array with positions enable for put disks)
        arrayCanReverse = this.canPutDisk(board, x, y);
        b[x][y].allowedCell = arrayCanReverse;
        if (arrayCanReverse.length > 0) cantAllow += 1;
        // count disks
        if (b[x][y].disk === 'white') disksWhite += 1;
        if (b[x][y].disk === 'black') disksBlack += 1;
      }
    }
    this.props.setAllowedCells(cantAllow); // dispatch how many cells enable
    this.props.addDisksPlayers({
      white: disksWhite, black: disksBlack,
    }); // dispatch how many disks for player
    return b;
  }

  // array with positions enable for put disks
  canPutDisk(board, x, y) {
    const b = board.slice();
    if (b[x][y].disk) return [];
    const canReverse = [];
    let X;
    let Y;
    let cantDisks;
    let cells;
    directions.forEach((direction) => {
      cantDisks = 0;
      cells = [];
      X = x;
      Y = y;
      do {
        cells.push({ X, Y });
        X += direction[0];
        Y += direction[1];
        cantDisks += 1;
      } while (this.inBoard(X, Y) && b[X][Y].disk === this.diskOponent());

      if (cantDisks > 1 && this.inBoard(X, Y) && b[X][Y].disk === this.props.currentPlayer) {
        canReverse.push(cells);
        // cells.forEach((cell) => {
        //   board[cell[0]][cell[1]].disk = this.props.currentPlayer;
        // });
      }
    });
    return [].concat(...canReverse);
  }

  async handleUndo() {
    await this.props.onUndo();
    // await this.props.changeTurn();
  }

  async handleRedo() {
    await this.props.onRedo();
    // await this.props.changeTurn();
  }

  render() {
    return (
      <>
        <BoardLayout board={this.props.board} />
        <button type="button" onClick={this.handleUndo.bind(this)} disabled={!this.props.canUndo}>
          Undo
        </button>
        <button type="button" onClick={this.handleRedo.bind(this)} disabled={!this.props.canRedo}>
          Redo
        </button>
      </>
    );
  }
}

const mapStateToProps = (state) => ({
  board: state.board.present.board,
  posDiskWhite: state.disks.present.posDisksWhite,
  posDiskBlack: state.disks.present.posDisksBlack,
  currentPlayer: state.players.currentPlayer,
  // Redux uno/redo
  canUndo: state.board.past.length > 1,
  canRedo: state.board.future.length > 0,
});

const mapDispatchToProps = {
  setBoard,
  setAllowedCells,
  addDisksPlayers,
  changeTurn,
  onUndo: () => UndoActionCreators.undo(),
  onRedo: () => UndoActionCreators.redo(),
};

export default connect(mapStateToProps, mapDispatchToProps)(Board);
