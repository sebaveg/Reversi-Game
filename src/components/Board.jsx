import React from 'react';

import Cell from '../containers/Cell';

import '../assets/styles/Board.css';

const Board = ({ board = [] }) => (
  <table className="Board">
    <tbody>
      {board.map((row, x) => (
        <tr key={x}>
          {board[x].map((cell, y) => (
            <Cell key={`${x}${y}`} disk={cell.disk} allowed={cell.allowedCell} position={[x, y]}>
              {cell.id}
            </Cell>
          ))}
        </tr>
      ))}
    </tbody>
  </table>
);
export default Board;
