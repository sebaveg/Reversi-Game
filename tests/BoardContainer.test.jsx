/* eslint-disable no-undef */
import React from 'react';
import configureMockStore from 'redux-mock-store';
import Board from '../src/containers/Board';

import BoardLayout from '../src/components/Board';

const initialState = {
  error: '',
  board: {
    past: [],
    present: {
      board: [
        [{ id: 'A1' }, { id: 'A2' }, { id: 'A3' }, { id: 'A4' }, { id: 'A5' }, { id: 'A6' }, { id: 'A7' }, { id: 'A8' }],
        [{ id: 'B1' }, { id: 'B2' }, { id: 'B3' }, { id: 'B4' }, { id: 'B5' }, { id: 'B6' }, { id: 'B7' }, { id: 'B8' }],
        [{ id: 'C1' }, { id: 'C2' }, { id: 'C3' }, { id: 'C4' }, { id: 'C5' }, { id: 'C6' }, { id: 'C7' }, { id: 'C8' }],
        [{ id: 'D1' }, { id: 'D2' }, { id: 'D3' }, { id: 'D4' }, { id: 'D5' }, { id: 'D6' }, { id: 'D7' }, { id: 'D8' }],
        [{ id: 'E1' }, { id: 'E2' }, { id: 'E3' }, { id: 'E4' }, { id: 'E5' }, { id: 'E6' }, { id: 'E7' }, { id: 'E8' }],
        [{ id: 'F1' }, { id: 'F2' }, { id: 'F3' }, { id: 'F4' }, { id: 'F5' }, { id: 'F6' }, { id: 'F7' }, { id: 'F8' }],
        [{ id: 'G1' }, { id: 'G2' }, { id: 'G3' }, { id: 'G4' }, { id: 'G5' }, { id: 'G6' }, { id: 'G7' }, { id: 'G8' }],
        [{ id: 'H1' }, { id: 'H2' }, { id: 'H3' }, { id: 'H4' }, { id: 'H5' }, { id: 'H6' }, { id: 'H7' }, { id: 'H8' }],
      ], // Array 2D
      currentPlayer: 'black',
    },
    future: [],
  },
  disks: {
    present: {
      posDisksWhite: [],
      posDisksBlack: [],
    },
  },
};

const mockStore = configureMockStore();

describe('<Board /> Container', () => {
  let wrapper;
  beforeEach(() => {
    const store = mockStore(initialState);
    wrapper = shallow(<Board store={store} />).childAt(0).dive();
  });

  it(' +++ capturing snapshot of home', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render one <div>', () => {
    const component = wrapper.find('div');
    expect(component.length).toBe(1);
  });

  it('should render one <button>', () => {
    const component = wrapper.find('button');
    expect(component.length).toBe(2);
  });

  it('should render one <BoardLayout /> component', () => {
    const component = wrapper.find(BoardLayout);
    expect(component.length).toBe(1);
  });

  it('+++ check prop matches with initialState', () => {
    // expect(container.prop('board')).toEqual(initialState.board.present);
    expect(wrapper.prop('posDiskWhite')).toEqual(initialState.posDiskWhite);
    expect(wrapper.prop('posDiskBlack')).toEqual(initialState.posDiskBlack);
    expect(wrapper.prop('currentPlayer')).toEqual(initialState.currentPlayer);
  });
});

// describe('<Board /> rendering', () => {
//   beforeEach(() => {
//     store = mockStore(initialState);
//     container = shallow(<Board store={store} />);
//   });
//   it('+++ <Board /> Container', () => {
//     expect(container.length).toEqual(1);
//   });
//   // });

// });

// describe('<Board /> rendering', () => {
//   beforeEach(() => {
//     store = mockStore(initialState);
//     component = mount(<Board />);
//   });
// });
