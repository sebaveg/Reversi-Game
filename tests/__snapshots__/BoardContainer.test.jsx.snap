// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> Container  +++ capturing snapshot of home 1`] = `
<Fragment>
  <Board
    board={
      Array [
        Array [
          Object {
            "id": "A1",
          },
          Object {
            "id": "A2",
          },
          Object {
            "id": "A3",
          },
          Object {
            "id": "A4",
          },
          Object {
            "id": "A5",
          },
          Object {
            "id": "A6",
          },
          Object {
            "id": "A7",
          },
          Object {
            "id": "A8",
          },
        ],
        Array [
          Object {
            "id": "B1",
          },
          Object {
            "id": "B2",
          },
          Object {
            "id": "B3",
          },
          Object {
            "id": "B4",
          },
          Object {
            "id": "B5",
          },
          Object {
            "id": "B6",
          },
          Object {
            "id": "B7",
          },
          Object {
            "id": "B8",
          },
        ],
        Array [
          Object {
            "id": "C1",
          },
          Object {
            "id": "C2",
          },
          Object {
            "id": "C3",
          },
          Object {
            "id": "C4",
          },
          Object {
            "id": "C5",
          },
          Object {
            "id": "C6",
          },
          Object {
            "id": "C7",
          },
          Object {
            "id": "C8",
          },
        ],
        Array [
          Object {
            "id": "D1",
          },
          Object {
            "id": "D2",
          },
          Object {
            "id": "D3",
          },
          Object {
            "id": "D4",
          },
          Object {
            "id": "D5",
          },
          Object {
            "id": "D6",
          },
          Object {
            "id": "D7",
          },
          Object {
            "id": "D8",
          },
        ],
        Array [
          Object {
            "id": "E1",
          },
          Object {
            "id": "E2",
          },
          Object {
            "id": "E3",
          },
          Object {
            "id": "E4",
          },
          Object {
            "id": "E5",
          },
          Object {
            "id": "E6",
          },
          Object {
            "id": "E7",
          },
          Object {
            "id": "E8",
          },
        ],
        Array [
          Object {
            "id": "F1",
          },
          Object {
            "id": "F2",
          },
          Object {
            "id": "F3",
          },
          Object {
            "id": "F4",
          },
          Object {
            "id": "F5",
          },
          Object {
            "id": "F6",
          },
          Object {
            "id": "F7",
          },
          Object {
            "id": "F8",
          },
        ],
        Array [
          Object {
            "id": "G1",
          },
          Object {
            "id": "G2",
          },
          Object {
            "id": "G3",
          },
          Object {
            "id": "G4",
          },
          Object {
            "id": "G5",
          },
          Object {
            "id": "G6",
          },
          Object {
            "id": "G7",
          },
          Object {
            "id": "G8",
          },
        ],
        Array [
          Object {
            "id": "H1",
          },
          Object {
            "id": "H2",
          },
          Object {
            "id": "H3",
          },
          Object {
            "id": "H4",
          },
          Object {
            "id": "H5",
          },
          Object {
            "id": "H6",
          },
          Object {
            "id": "H7",
          },
          Object {
            "id": "H8",
          },
        ],
      ]
    }
  />
  <div
    className="wrapper-btn-undo-redo"
  >
    <button
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Undo
    </button>
    <button
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Redo
    </button>
  </div>
</Fragment>
`;
